A channelz sampler

mohanli-ml opened the pull request:
A channelz sampler that shows the snapshoted structure of servers and channels of a gRPC process.
The channelz sampler take two flags: server_address, custom_credentials_type




@apolcyn

mohanli-ml commented:
I signed CLA

mohanli-ml commented:
> [![CLA Check](https://camo.githubusercontent.com/15ec12d849d4e413c7c229c0c532992ccabacda4/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f747369676e65642e706e67)](https://api.lfcla.com/v2/repository-provider/github/sign/1656153/27729880/23396)
> 
> * [❌](https://api.lfcla.com/v2/repository-provider/github/sign/1656153/27729880/23396)  mohanli-ml The commit ([2b37810](https://github.com/grpc/grpc/commit/2b3781097487531fd897b8526c8b4d8d5e0c0dba)) is not authorized under a signed CLA. [Please click here to be authorized](https://api.lfcla.com/v2/repository-provider/github/sign/1656153/27729880/23396). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

I signed it

mohanli-ml commented:
I signed it

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+
+DEFINE_string(server_address, "", "channelz server address");
+DEFINE_string(custom_credentials_type, "", "custom credentials type");
+// "INSECURE_CREDENTIALS" - InsecureChannelCredentials
```
nit: unnecessary comment, can we delete it?

If this is meant to be a kind of help message, then it should be included in the flag'd description.

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+int main(int argc, char** argv) {
+  // make sure flags can be used
+  grpc::testing::InitTest(&argc, &argv, true);
+  std::cout << "server address: " << FLAGS_server_address << std::endl;
```
nit: lets remove printing of flags back to stdout -- I think it will likely clutter the output without too much benefit here

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+            << std::endl;
+
+  // create a channelz client
+  ::grpc::channelz::experimental::InitChannelzService();
```
Calling `::grpc::channelz::experimental::InitChannelzService();` is unnecessary here, no? There are no server endpoints running in this tool's process?

mohanli-ml replied:
Yes for a channelz client this is not necessary. I have removed it.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+using grpc::ClientContext;
+using grpc::Status;
+
+int64_t GetServerID(grpc::channelz::v1::Server& server) {
```
Suggest inlining the contents of these one-liner helpers, which I think would be slightly more readable.

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+  // Get all servers by calling GetServers
+  GetServersRequest get_server_request;
+  GetServersResponse get_server_response;
+  ClientContext get_server_context;
```
It's not safe to reuse the same `ClientContext` object across multiple RPCs (also note https://github.com/grpc/grpc/blob/3201d7f7bfca6e6d3883da710c105d3e8167e0a3/include/grpcpp/impl/codegen/client_context_impl.h#L181).

We should restructure things to ensure that the same we use a new/fresh `ClientContext` object on each RPC, e.g. in the `GetServers` loop below.

mohanli-ml replied:
Thanks for pointing out. I have put the `ClientContent` into the `while` loop.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+    Status status = channelz_stub->GetServers(
+        &get_server_context, get_server_request, &get_server_response);
+    if (!status.ok()) {
+      std::cout << "Channelz failed: " << status.error_message() << std::endl;
```
for warning/error logs, lets use `gpr_log(GPR_ERROR, ....)`, instead of printing to stdout.

In general, lets reserve stdout for only the channelz dump output.

mohanli-ml replied:
done.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+        &get_server_context, get_server_request, &get_server_response);
+    if (!status.ok()) {
+      std::cout << "Channelz failed: " << status.error_message() << std::endl;
+      return 1;
```
nit: lets abort rather than return 1

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+    Status status = channelz_stub->GetServers(
+        &get_server_context, get_server_request, &get_server_response);
+    if (!status.ok()) {
+      std::cout << "Channelz failed: " << status.error_message() << std::endl;
```
lets include the context's `debug_error_string` in the error log message -- which has richer debug info.

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+
+  // Server side code
+  // Get all servers by calling GetServers
+  GetServersRequest get_server_request;
```
nit: `GetServersRequest` and `GetServersResponse` can be loop-local

mohanli-ml replied:
done.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+  std::cout << "Number of servers = " << top_servers.size() << std::endl;
+
+  // print out sockets of each server
+  for (auto& i : top_servers) {
```
nit: can we rename `i` to `server`, and can we make this loop variable const?

mohanli-ml replied:
Will not write such ugly style code in the future...

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+  // print out sockets of each server
+  for (auto& i : top_servers) {
+    std::cout << "Server " << GetServerID(i) << " listen_socket: ";
+    for (auto& j : i.listen_socket()) {
```
nit: can we rename `j` to `socket`, and make this const too? (j and i make me think they are index variables)

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+  for (auto& i : top_servers) {
+    std::cout << "Server " << GetServerID(i) << " listen_socket: ";
+    for (auto& j : i.listen_socket()) {
+      std::cout << j.socket_id() << " ";
```
It would be more interesting for debug purposes to see the entire `Server` data structure.

Can we use the protobuf text format APIs to dump the entire `Server` objects (https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format)

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+  // Get all top channels by calling GetTopChannels
+  GetTopChannelsRequest get_top_channels_request;
+  GetTopChannelsResponse get_top_channels_response;
+  ClientContext get_top_channels_context;
```
similar to above, we'll need to make sure this context object isn't reused across RPCs

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+
+  // Client side code
+  // Get all top channels by calling GetTopChannels
+  GetTopChannelsRequest get_top_channels_request;
```
this request/response objects can also be local to the loop they're used in

mohanli-ml replied:
Done. Will keep request/response/content local in the future.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+                                                  get_top_channels_request,
+                                                  &get_top_channels_response);
+    if (!status.ok()) {
+      std::cout << "Channelz failed: " << status.error_message() << std::endl;
```
same error handling here as suggested above for the server requests

mohanli-ml replied:
done

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+    }
+  }
+  std::cout << "Number of top channels = " << top_channels.size() << std::endl;
+
```
nit: lets remove unnecessary blank lines

mohanli-ml replied:
remove blank line on line 279

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+    std::cout << "channel: ";
+    for (auto& i : subchannel.channel_ref()) {
+      int64_t ch_id = i.channel_id();
+      std::cout << ch_id << " ";
```
similar to above, it would be more interesting to dump the full proto messages, rather than just the channel/subchannel/socket id's

mohanli-ml replied:
Keep the current output format for quickly showing the tree structure, and print out entity (i.e., channel, subchannel, server, socket) data in the end.

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,289 @@
+    std::cout << std::endl;
+  }
+
+  // store result
```
nit: please remove stale comments

mohanli-ml replied:
done

mohanli-ml commented:
The update in commit f6e08ff:
1. Wrap channelz_sampler to a class;
2. Add a feature of dumping data of all entities (channel, subchannel, server, socket) to stdout;
3. Add a test file for testing the channelz_sampler. The test implements a simple echo RPC services with one server and two clients. 

An example output of the test:
Number of servers = 1
Server 1 listen_socket: 2 
Number of top channels = 2
Tree depth = 0
    Channel 3 descendence - subchannel: 6 
Tree depth = 1
    Subchannel 6 descendence - socket: 8 
Tree depth = 0
    Channel 4 descendence - subchannel: 6 
Tree depth = 1
    Subchannel 6 descendence - socket: 8 
channel 3 data:
state {
  state: READY
}
target: "0.0.0.0:10000"
trace {
  num_events_logged: 6
  creation_timestamp {
    seconds: 1594423368
    nanos: 394812677
  }
  events {
    description: "Channel created"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 394812687
    }
  }
  events {
    description: "Channel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812725
    }
  }
  events {
    description: "Created new LB policy \"pick_first\""
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812709
    }
  }
  events {
    description: "Resolution event: Address list became non-empty"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812707
    }
  }
  events {
    description: "Channel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812715
    }
  }
  events {
    description: "Channel state change to READY"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 397812717
    }
  }
}
calls_started: 1
calls_failed: 1
last_call_started_timestamp {
  seconds: 1594423368
  nanos: 395943891
}
channel 4 data:
state {
  state: READY
}
target: "0.0.0.0:10000"
trace {
  num_events_logged: 6
  creation_timestamp {
    seconds: 1594423368
    nanos: 394812720
  }
  events {
    description: "Channel created"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 394812680
    }
  }
  events {
    description: "Channel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812716
    }
  }
  events {
    description: "Created new LB policy \"pick_first\""
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812709
    }
  }
  events {
    description: "Resolution event: Address list became non-empty"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812728
    }
  }
  events {
    description: "Channel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812711
    }
  }
  events {
    description: "Channel state change to READY"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 397812721
    }
  }
}
calls_started: 1
calls_failed: 1
last_call_started_timestamp {
  seconds: 1594423368
  nanos: 395945411
}
subchannel 6 data:
state {
  state: READY
}
target: "ipv4:0.0.0.0:10000"
trace {
  num_events_logged: 3
  creation_timestamp {
    seconds: 1594423368
    nanos: 395812696
  }
  events {
    description: "subchannel created"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812719
    }
  }
  events {
    description: "Subchannel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812716
    }
  }
  events {
    description: "Subchannel state change to READY"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 397812706
    }
  }
}
calls_started: 2
calls_failed: 2
last_call_started_timestamp {
  seconds: 1594423368
  nanos: 398814620
}


subchannel 6 data:
state {
  state: READY
}
target: "ipv4:0.0.0.0:10000"
trace {
  num_events_logged: 3
  creation_timestamp {
    seconds: 1594423368
    nanos: 395812696
  }
  events {
    description: "subchannel created"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812719
    }
  }
  events {
    description: "Subchannel state change to CONNECTING"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 395812716
    }
  }
  events {
    description: "Subchannel state change to READY"
    severity: CT_INFO
    timestamp {
      seconds: 1594423368
      nanos: 397812706
    }
  }
}
calls_started: 2
calls_failed: 2
last_call_started_timestamp {
  seconds: 1594423368
  nanos: 398814610
}
server 1 data:
trace {
  num_events_logged: 1
  creation_timestamp {
    seconds: 1594423365
    nanos: 400812720
  }
  events {
    description: "Server created"
    severity: CT_INFO
    timestamp {
      seconds: 1594423365
      nanos: 400812729
    }
  }
}
calls_started: 3
calls_failed: 2
last_call_started_timestamp {
  seconds: 1594423368
  nanos: 611967329
}
server 2 data:
server 8 data:
streams_started: 2
streams_failed: 2
messages_sent: 2
last_local_stream_created_timestamp {
  seconds: 1594423368
  nanos: 398871161
}
last_message_sent_timestamp {
  seconds: 1594423368
  nanos: 546282921
}
server 8 data:
streams_started: 2
streams_failed: 2
messages_sent: 2
last_local_stream_created_timestamp {
  seconds: 1594423368
  nanos: 398871137
}
last_message_sent_timestamp {
  seconds: 1594423368
  nanos: 546282907
}

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,167 @@
+  std::thread client_thread_1(RunClient, "1");
+  std::thread client_thread_2(RunClient, "2");
+  float run_services_seconds = 5.0;
+  std::cout << "Run echo service for " << run_services_seconds << " seconds."
```
nit: please use `gpr_log` in general, rather than printing to stdout or stderr

mohanli-ml replied:
All stdout and stderr in the test have been replace by `gpr_log` .

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,167 @@
+  sleep(server_start_seconds);
+
+  // client thread
+  std::thread client_thread_1(RunClient, "1");
```
please ensure to explicitly join all threads created

mohanli-ml replied:
Done

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,167 @@
+  // Run the channelz sampler
+  std::string channelz_sampler_bin_path =
+      "./bazel-bin/test/cpp/util/channelz_sampler";
+  grpc::SubProcess* test_driver = new grpc::SubProcess(
```
we'll need to wait for this process's completion and make sure the status (e.g. as is done in https://github.com/grpc/grpc/blob/6ff6066dd59aff38fec77a25127427609983b650/test/cpp/naming/resolver_component_tests_runner_invoker.cc#L130)

mohanli-ml replied:
Done

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,167 @@
+  float run_services_seconds = 5.0;
+  std::cout << "Run echo service for " << run_services_seconds << " seconds."
+            << std::endl;
+  sleep(run_services_seconds);
```
what's the reason for sleeping here? Can we remove this?

mohanli-ml replied:
Removed

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,167 @@
+      {std::move(channelz_sampler_bin_path),
+       "--server_address=" + server_address,
+       "--custom_credentials_type=" + custom_credentials_type});
+  while (true)
```
This loop will need to be removed

mohanli-ml replied:
Removed

apolcyn commented on test/cpp/util/channelz_sampler.cc:
```diff
@@ -0,0 +1,377 @@
+
+  // Server side code
+  bool isRPCOK = channelz_sampler.GetServersRPC();
+  if (!isRPCOK) abort();
```
nit: lets use `GPR_ASSERT(0)` (which will print source line info) rather than vanilla abort

mohanli-ml replied:
Done

mohanli-ml commented:
New features:
1. Addressed the comments about the test;
2. Wrapped the channelz_sampler into a class to make the code more readable;
3. Accept three new parameters: sampling_times, sampling_interval_seconds, output_json;
4. Add the feature to store a Json output;
5. Add a python script, which draws a Gantt style chart based on the Json output.

An output example is as follows, which samples twice and the sampling interval is 5s:

Wait for sampling interval 5s...
##### 0th sampling #####
Number of servers = 1
Server 1 listen_socket: 2 
Number of top channels = 0
server 1 data:
trace {
  num_events_logged: 1
  creation_timestamp {
    seconds: 1593703162
    nanos: 856305418
  }
  events {
    description: "Server created"
    severity: CT_INFO
    timestamp {
      seconds: 1593703162
      nanos: 856305441
    }
  }
}
calls_started: 170
calls_succeeded: 169
last_call_started_timestamp {
  seconds: 1594660382
  nanos: 395358167
}
socket 2 data:
Wait for sampling interval 5s...
##### 1th sampling #####
Number of servers = 1
Server 1 listen_socket: 2 
Number of top channels = 0
server 1 data:
trace {
  num_events_logged: 1
  creation_timestamp {
    seconds: 1593703162
    nanos: 856305418
  }
  events {
    description: "Server created"
    severity: CT_INFO
    timestamp {
      seconds: 1593703162
      nanos: 856305441
    }
  }
}
calls_started: 173
calls_succeeded: 172
last_call_started_timestamp {
  seconds: 1594660387
  nanos: 399685766
}
socket 2 data:

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,172 @@
+  std::thread server_thread(RunServer);
+  gpr_log(GPR_INFO, "Wait 3 seconds to make sure server is started...");
+  float server_start_seconds = 3.0;
+  sleep(server_start_seconds);
```
instead of sleeping for 3 seconds, can we create a channel that tries to connect to the server, and returns when it does connect, using the `WaitForConnected` channel API?

E.g., https://github.com/grpc/grpc/blob/master/test/cpp/end2end/grpclb_end2end_test.cc#L959

mohanli-ml replied:
Done. Thanks for pointing out.

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,172 @@
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
```
we'll need to use some signal to know when to terminate the loop upon cleanup/thread-join

Suggest using `gpr_event`. See e.g. https://github.com/grpc/grpc/blob/8febefc05784fe2e12a5933a96d373297ad0cfdf/test/cpp/naming/resolver_component_test.cc#L312

mohanli-ml replied:
Done

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,172 @@
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
+    Status status = echoer.EmptyCall();
```
nit:  no need for `status` return value here if we're ignoring it

mohanli-ml replied:
Removed status return value.

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
```
nit: I don't see a lot of advantage to this being a class, so I suggest removing this, and inlining the fields and logic of this class into the `RunClient` loop.

mohanli-ml replied:
Removed

apolcyn commented on test/cpp/util/channelz_sampler_test.cc:
```diff
@@ -0,0 +1,172 @@
+  client_thread_1.join();
+  client_thread_2.join();
+
+  EXPECT_TRUE(true);
```
this is a no-op, right? can we remove this?

mohanli-ml replied:
Removed
